From 24bd7a0b5ae1e480130e33d7478f823f0a86a109 Mon Sep 17 00:00:00 2001
From: bowehan <bowehan@qti.qualcomm.com>
Date: Mon, 11 Aug 2025 14:45:41 +0800
Subject: [PATCH 04/12] FROMLIST: drm/msm/dp: Move link training to
 atomic_enable()

Currently, the DP link training is being done during HPD. Move link training to msm_dp_display_prepare() in accordance with the msm_dp_display_prepare()

Link disabling is already done in atomic_post_disable() (as part of the
dp_ctrl_off_link_stream() helper).

Finally, call the plug/unplug handlers directly in hpd_notify() instead
of queueing them in the event thread so that they aren't preempted by
other events.

Signed-off-by: Jessica Zhang <jessica.zhang@oss.qualcomm.com>
Signed-off-by: bowehan <bowehan@qti.qualcomm.com>
Link: https://lore.kernel.org/all/20250808-hpd-refactor-v2-4-7f4e1e741aa3@oss.qualcomm.com/
Upstream-Status: Submitted [https://lore.kernel.org/all/20250808-hpd-refactor-v2-4-7f4e1e741aa3@oss.qualcomm.com/]
---
 drivers/gpu/drm/msm/dp/dp_display.c | 15 +++++++++------
 1 file changed, 9 insertions(+), 6 deletions(-)

diff --git a/drivers/gpu/drm/msm/dp/dp_display.c b/drivers/gpu/drm/msm/dp/dp_display.c
index 917dcfb1fa0f..dd38dfa10b4c 100644
--- a/drivers/gpu/drm/msm/dp/dp_display.c
+++ b/drivers/gpu/drm/msm/dp/dp_display.c
@@ -507,11 +507,6 @@ static int msm_dp_display_process_hpd_high(struct msm_dp_display_private *dp)
 		msm_dp_display_mst_init(dp);
 
 	msm_dp_link_reset_phy_params_vx_px(dp->link);
-	rc = msm_dp_ctrl_on_link(dp->ctrl, dp_display->mst_active);
-	if (rc) {
-		DRM_ERROR("failed to complete DP link training\n");
-		goto end;
-	}
 
 	msm_dp_display_set_mst_mgr_state(dp, true);
 
@@ -1006,7 +1001,15 @@ static int msm_dp_display_prepare(struct msm_dp_display_private *dp)
 
 	if (dp->hpd_state == ST_CONNECTED && !dp->active_stream_cnt) {
 		msm_dp_display_host_phy_init(dp);
-		force_link_train = true;
+	}
+
+	if (!dp->active_stream_cnt) {
+		rc = msm_dp_ctrl_on_link(dp->ctrl, msm_dp_display->mst_active);
+		if (rc) {
+			DRM_ERROR("Failed link training (rc=%d)\n", rc);
+			msm_dp_display->connector->state->link_status = DRM_LINK_STATUS_BAD;
+			force_link_train = true;
+		}
 	}
 
 	rc = msm_dp_ctrl_prepare_stream_on(dp->ctrl, force_link_train);
-- 
2.34.1

